tensorflow/stream_executor/cuda/cuda_blas.cc

```diff
@@ -1762,14 +1846,26 @@ bool CUDABlas::DoBlasGemm(
                       "precondition violation";
     }
   }
-  // TODO(sesse): Consider supporting the Hgemm interface, which uses half
-  // calculations internally (faster on newer devices, such as Pascal and TX1,
-  // but less precise).
-  return DoBlasInternal(
+
+  bool use_tensor_ops = false;
+#if CUDA_VERSION >= 9000
+  int cc_major, cc_minor;
+  stream->parent()->GetDeviceDescription().cuda_compute_capability(&cc_major,
+                                                                   &cc_minor);
+
+  // GPUs < sm_70 don't support tensor cores
+  if (cc_major >= 7 && TensorOpMathEnabled()) {
+    use_tensor_ops = true;
+  }
+#endif
+
+  return DoBlasInternalWithTensorOps(
       wrap::cublasSgemmEx, stream, true /* = pointer_mode_host */,
-      CUDABlasTranspose(transa), CUDABlasTranspose(transb), m, n, k, &alpha,
-      CUDAMemory(a), SE_CUDA_DATA_HALF, lda, CUDAMemory(b), SE_CUDA_DATA_HALF,
-      ldb, &beta, CUDAMemoryMutable(c), SE_CUDA_DATA_HALF, ldc);
+      use_tensor_ops, CUDABlasTranspose(transa), CUDABlasTranspose(transb), m,
```

Comment:
DoBlasInternalWithTensorOps should set use_tensor_ops to be true, I think we should get rid of use_tensor_ops argument.

Reply:
It is convenient to pass this as a flag, otherwise I'll have to do an if-else on the flag in order to make different function calls. Alternatively, I could remove DoBlasInternalWithTensorOps entirely and use direct calls to DoBlasInternalImpl. Would that be preferable?

Reply:
I'm fine with either way. I noticed that later in DoBlasInternalWithTensorOps() you directly called this: return DoBlasInternalImpl(cublas_func, stream, pointer_mode_host, /*err_on_failure=*/true, /*use_tensor_ops=*/false, ...) Shouldn't it be /*use_tensor_ops=*/true?
Also I still think if you are using DoBlasInternalWithTensorOps(), it suggests you are using tensor ops, so probably just get rid of this function?

Reply:
Yes, that was a bug. Thanks for the catch. DoBlasGemm for fp16 now calls  DoBlasInternalImpl directly and DoBlasInternalWithTensorOps has been removed.